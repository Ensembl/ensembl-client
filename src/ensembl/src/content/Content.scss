@import 'src/styles/common';

/*
 TODO:
  Add a global layout component based on CSS grid:

  | HEADER  |
  | CONTENT |

  so that the area allotted to content is calculated automatically by the browser,
  and not specifically defined as currently is done below
*/

.content {
  &.shorter {
    height: calc(100% - 79px);
  }

  &.taller {
    height: calc(100% - 25px);
  }
}
