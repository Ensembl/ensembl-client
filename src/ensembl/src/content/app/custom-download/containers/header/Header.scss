@import 'src/styles/common';

.wrapper {
  background-color: $ens-light-grey;
  height: 60px;
  position: relative;
}

.resultCounter {
  position: absolute;
  top: 50%;
  left: 50px;
  color: $ens-black;
  margin-top: -10px;
}

.selectedFilters {
  position: absolute;
  top: 50%;
  left: 40%;
  margin-top: -12px;
}

.previewButton {
  position: absolute;
  right: 30px;
  top: 50%;
  margin-top: -15px;
}

/* RoundButton extended styles*/
.active {
  background-color: $ens-medium-dark-grey;
  border: 1px solid $ens-medium-dark-grey;
  color: $ens-white;
  margin-right: 10px;
}

.previewResultCounter {
  margin-left: 20px;
  padding-top: 12px;
  display: inline-block;
}

.boldResultCounter {
  font-weight: bold;
  margin-right: 2px;
}

.resultsLabel {
  font-size: 11px;
  color: $ens-dark-grey;
}
.saveConfiguration {
  font-size: 11px;
  color: $ens-blue;
}

.backButton {
  height: 30px;
  width: 30px;
  margin-left: 20px;
  display: inline-block;
}

.downloadTypeSelectHolder {
  display: inline-block;
  margin-left: 20px;
}

.downloadTypeLabel {
  color: $ens-dark-grey;
}

.downloadTypeSelect {
  margin-left: 10px;
  color: $ens-blue;
  li {
    color: $ens-medium-dark-grey;
  }
}
