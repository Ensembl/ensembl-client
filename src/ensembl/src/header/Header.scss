@import 'src/styles/common';

.topBar {
  @include top-bar-container;
  @include top-bar-unstack;

  color: $white;
  padding: 2px 20px;
}

.homeLink {
  display: inline-block;
  font-size: 14px;
  font-weight: $global-weight-bold;

  a {
    color: $ens-white;
  }

  img {
    bottom: 2px;
    height: 16px;
    margin-right: 2px;
    position: relative;
    width: 16px;
  }
}

.strapline,
.copyright {
  display: none;
  font-size: 11px;
  margin-left: 30px;

  @include breakpoint(medium) {
    display: inline-block;
  }

  a {
    color: $ens-blue;
  }
}
