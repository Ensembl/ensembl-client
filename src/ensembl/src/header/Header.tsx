import React, { memo, FunctionComponent } from 'react';
import { Link } from 'react-router-dom';

import HeaderButtons from './header-buttons/HeaderButtons';
import LaunchbarContainer from './launchbar/LaunchbarContainer';
import Account from './account/Account';

import homeIcon from 'static/img/header/home.svg';

import styles from './Header.scss';

type HeaderProps = {};

export const HomeLink = () => (
  <div className={styles.homeLink}>
    <Link to="/">
      <img src={homeIcon} alt="" /> Ensembl
    </Link>
  </div>
);

export const ReleaseVersion = () => (
  <div className={styles.strapline}>Pre-release - March 2019</div>
);

export const Copyright = () => (
  <div className={styles.copyright}>
    <a href="https://www.ebi.ac.uk" target="_blank" rel="noopener noreferrer">
      Â© EMBL-EBI
    </a>
  </div>
);

export const Header: FunctionComponent<HeaderProps> = () => (
  <header>
    <div className={styles.topBar}>
      <div>
        <HomeLink />
        <ReleaseVersion />
        <Copyright />
      </div>
      <HeaderButtons />
    </div>
    <Account />
    <LaunchbarContainer />
  </header>
);

export default memo(Header);
