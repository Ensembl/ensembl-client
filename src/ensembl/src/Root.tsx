import React, {
  FunctionComponent,
  useEffect,
  useState,
  useCallback
} from 'react';
import { BrowserRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { withCookies, ReactCookieProps, Cookies } from 'react-cookie';
import useResizeObserver from 'use-resize-observer';

import Header from './header/Header';
import Content from './content/Content';
import PrivacyBanner from './shared/privacy-banner/PrivacyBanner';

import { updateBreakpointWidth } from './globalActions';
import { getBreakpointWidth } from './globalSelectors';
import { RootState } from './rootReducer';
import { BreakpointWidth, getBreakpoint } from './globalConfig';

import styles from './Root.scss';

type StateProps = {
  breakpointWidth: BreakpointWidth;
};

type DispatchProps = {
  updateBreakpointWidth: (breakpointWidth: BreakpointWidth) => void;
};

type OwnProps = {};

type RootProps = StateProps & DispatchProps & ReactCookieProps & OwnProps;

const Root: FunctionComponent<RootProps> = (props: RootProps) => {
  const [ref, width] = useResizeObserver();
  const currentBreakpoint: BreakpointWidth = getBreakpoint(width);

  useEffect(() => {
    props.updateBreakpointWidth(currentBreakpoint);
  }, [props.updateBreakpointWidth, currentBreakpoint]);

  const [showPrivacyBanner, setShowPrivacyBanner] = useState(true);
  const cookies = props.cookies as Cookies;

  useEffect(() => {
    if (cookies.get('ENSEMBL_PRIVACY_POLICY') === 'true') {
      setShowPrivacyBanner(false);
    }
  }, [cookies]);

  const closeBanner = useCallback(() => {
    cookies.set('ENSEMBL_PRIVACY_POLICY', 'true');
    setShowPrivacyBanner(false);
  }, [cookies]);

  return (
    <BrowserRouter>
      <div ref={ref as React.RefObject<HTMLDivElement>} className={styles.root}>
        <Header />
        <Content />
        {showPrivacyBanner ? (
          <PrivacyBanner closeBanner={closeBanner} />
        ) : (
          false
        )}
      </div>
    </BrowserRouter>
  );
};

const mapStateToProps = (state: RootState): StateProps => ({
  breakpointWidth: getBreakpointWidth(state)
});

const mapDispatchToProps: DispatchProps = { updateBreakpointWidth };

export default withCookies(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Root)
);
