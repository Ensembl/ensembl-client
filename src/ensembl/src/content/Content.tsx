import React, { FunctionComponent } from 'react';
import { Route, withRouter } from 'react-router-dom';
import { RouteComponentProps } from 'react-router';
import { connect } from 'react-redux';

import { RootState } from '../rootReducer';
import { getLaunchbarExpanded } from '../header/headerSelectors';

import Home from './home/Home';
import App from './app/App';

import styles from './Content.scss';

type StateProps = {
  launchbarExpanded: boolean;
};

type OwnProps = {};

type ContentProps = RouteComponentProps & StateProps & OwnProps;

export const getHeightClass = (launchbarExpanded: boolean): string => {
  return launchbarExpanded ? styles.shorter : styles.taller;
};

export const Content: FunctionComponent<ContentProps> = (
  props: ContentProps
) => {
  return (
    <main
      className={`${styles.content} ${getHeightClass(props.launchbarExpanded)}`}
    >
      <Route path="/" component={Home} exact={true} />
      <Route path="/app" component={App} />
    </main>
  );
};

const mapStateToProps = (state: RootState): StateProps => ({
  launchbarExpanded: getLaunchbarExpanded(state)
});

export default withRouter(connect(mapStateToProps)(Content));
