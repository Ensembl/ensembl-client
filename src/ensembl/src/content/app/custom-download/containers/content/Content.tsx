import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import styles from './Content.scss';
import {
  getSelectedTabButton,
  getShowPreviewResult,
  getPreviewResult
} from '../../customDownloadSelectors';
import { setAttributes } from './attributes-accordion/attributesAccordionActions';
import { getAttributes } from './attributes-accordion/attributesAccordionSelector';

import AttributesAccordion from './attributes-accordion/AttributesAccordion';
import FiltersAccordion from './filter-accordion/FiltersAccordion';
import TabButtons from './tab-buttons/TabButtons';
import ResultHolder from './result-holder/ResultHolder';
import { RootState } from 'src/store';
import { attributes } from '../../sampledata';
import PreviewTable from '../../components/preview-table/PreviewTable';
// import { fetchGeneAttributes } from 'src/services/custom-download';

import { getSelectedAttributes, formatResults } from './result-holder/helpers';

type Props = StateProps & DispatchProps;

// const getGeneAttributes = async (props: Props) => {
//   const allAttributes = await fetchGeneAttributes();

//   const attributes: any = {};

//   let sections = ['gene'];

//   allAttributes.forEach((attribute: any) => {
//     sections = ['gene'];
//     if (attribute.type === 'TERM') {
//       if (attribute.name.split('.').length > 1) {
//         sections = attribute.name.split('.');
//       }
//     }

//     if (!attributes[sections[0]]) {
//       attributes[sections[0]] = {};
//     }
//     if (sections[0] && sections.length !== 1 && !attributes[sections[0]]) {
//       attributes[sections[0]] = {};
//     } else if (sections.length == 1) {
//       if(!attributes[sections[0]]['default']){
//         attributes[sections[0]]['default'] = {};
//       }
//       attributes[sections[0]]['default'][attribute.name] = {
//         id: attribute.name,
//         label: attribute.displayName,
//         checkedStatus: false
//       };
//     }

//     if (
//       sections[1] &&
//       sections.length !== 2 &&
//       !attributes[sections[0]][sections[1]]
//     ) {
//       attributes[sections[0]][sections[1]] = {};
//     } else if (sections.length == 2) {
//       if(!attributes[sections[0]]['default']){
//         attributes[sections[0]]['default'] = {};
//       }
//       attributes[sections[0]]['default'][sections[1]] = {
//         id: attribute.name,
//         label: attribute.displayName,
//         checkedStatus: false
//       };
//     }

//     if (
//       sections[2] &&
//       sections.length !== 3 &&
//       !attributes[sections[0]][sections[1]][sections[2]]
//     ) {
//       attributes[sections[0]][sections[1]][sections[2]] = {};
//     } else if (sections.length == 3) {
//       if(!attributes[sections[0]][sections[1]]['default']){
//         attributes[sections[0]][sections[1]]['default'] = {};
//       }
//       attributes[sections[0]][sections[1]]['default'][sections[2]] = {
//         id: attribute.name,
//         label: attribute.displayName,
//         checkedStatus: false
//       };
//     }

//     if (
//       sections[3] &&
//       sections.length !== 4 &&
//       !attributes[sections[0]][sections[1]][sections[2]][sections[3]]
//     ) {
//       attributes[sections[0]][sections[1]][sections[2]][sections[3]] = {};
//     } else if (sections.length == 4) {

//       if(!attributes[sections[0]][sections[1]][sections[2]]['default']){
//         attributes[sections[0]][sections[1]][sections[2]]['default'] = {};
//       }

//       attributes[sections[0]][sections[1]][sections[2]]['default'][sections[3]] = {
//         id: attribute.name,
//         label: attribute.displayName,
//         checkedStatus: false
//       };
//     }
//   });
//   props.setAttributes(attributes);
// };

const Content = (props: Props) => {
  useEffect(() => {
    // getGeneAttributes(props);
    props.setAttributes(attributes);
  }, []);

  const selectedAttributes: any = getSelectedAttributes(props.attributes);
  let formatedPreviewResult = [];
  if (props.previewResult.results) {
    formatedPreviewResult = formatResults(
      props.previewResult,
      selectedAttributes
    );
  }

  return (
    <div>
      {!props.showPreview && (
        <div>
          <div className={styles.resultList}>
            <ResultHolder />
          </div>
          <div className={styles.tabList}>
            <TabButtons />
          </div>
          <div className={styles.dataSelector}>
            {props.selectedTabButton === 'attributes' && (
              <AttributesAccordion />
            )}
            {props.selectedTabButton === 'filter' && <FiltersAccordion />}
          </div>
        </div>
      )}
      {props.showPreview && (
        <div>
          {' '}
          <PreviewTable resultData={formatedPreviewResult} />{' '}
        </div>
      )}
    </div>
  );
};

type StateProps = {
  selectedTabButton: string;
  showPreview: boolean;
  previewResult: any;
  attributes: any;
};

const mapStateToProps = (state: RootState): StateProps => ({
  selectedTabButton: getSelectedTabButton(state),
  showPreview: getShowPreviewResult(state),
  previewResult: getPreviewResult(state),
  attributes: getAttributes(state)
});

type DispatchProps = {
  setAttributes: (setAttributes: {}) => void;
};

const mapDispatchToProps: DispatchProps = {
  setAttributes
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Content);
