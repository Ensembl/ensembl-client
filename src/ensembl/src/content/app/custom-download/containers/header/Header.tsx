import React, { useCallback } from 'react';
import { connect } from 'react-redux';
import { SecondaryButton, PrimaryButton } from 'src/shared/button/Button';

import RoundButton, {
  RoundButtonStatus
} from 'src/shared/round-button/RoundButton';

import { RootState } from 'src/store';
import Select, { Option } from 'src/shared/select/Select';

import {
  getSelectedPreFilter,
  getPreviewResult,
  getShowPreviewResult,
  getDownloadType
} from '../../customDownloadSelectors';

import { getFilters } from '../content/filter-accordion/filterAccordionSelector';
import { getAttributes } from '../content/attributes-accordion/attributesAccordionSelector';

import {
  togglePreFiltersPanel,
  setShowPreview,
  setDownloadType
} from '../../customDownloadActions';
import ImageButton from 'src/shared/image-button/ImageButton';
import { ReactComponent as BackIcon } from 'static/img/shared/chevron-left.svg';

import {
  getSelectedAttributes,
  getSelectedFilters
} from '../content/result-holder/helpers';

import { fetchCustomDownloadResults } from 'src/services/custom-download';

import styles from './Header.scss';

type Props = StateProps & DispatchProps;

const downloadTypeSelectOptions = [
  {
    label: 'CSV',
    value: 'text/csv',
    isSelected: false
  },
  {
    label: 'JSON',
    value: 'application/json',
    isSelected: false
  }
  // {
  //   label: 'CSV.gz',
  //   value: 'text/csv',
  //   isSelected: false
  // },
  // {
  //   label: 'HTML',
  //   value: 'application/json',
  //   isSelected: false
  // },
  // {
  //   label: 'TSV',
  //   value: 'application/json',
  //   isSelected: false
  // },
  // {
  //   label: 'TSV.gz',
  //   value: 'application/json',
  //   isSelected: false
  // },
  // {
  //   label: 'URL',
  //   value: 'application/json',
  //   isSelected: false
  // },
  // {
  //   label: 'XML',
  //   value: 'application/xml',
  //   isSelected: false
  // },
  // {
  //   label: 'XLS',
  //   value: 'application/xml',
  //   isSelected: false
  // },
  // {
  //   label: 'XLS.gz',
  //   value: 'application/xml',
  //   isSelected: false
  // }
];

const getFormattedTotal = (total: number) => {
  if (!total) {
    return 0;
  }
  return total.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
};

const Header = (props: Props) => {
  const filterOnClick = () => {
    props.togglePreFiltersPanel(true);
  };

  const previewButtonOnClick = useCallback(() => {
    props.setShowPreview(true);
  }, [props.showPreview]);

  const showFilters = useCallback(() => {
    props.setShowPreview(false);
  }, [props.showPreview]);

  const resultCount = getFormattedTotal(props.previewResult.resultCount);

  const handleDownloadTypeSelect = useCallback(
    (option: string) => {
      props.setDownloadType(option);
    },
    [props.downloadType]
  );

  const selectOptions = [...downloadTypeSelectOptions].map((option: Option) => {
    const optionClone = { ...option };
    if (optionClone.value === props.downloadType) {
      optionClone.isSelected = true;
    }
    return optionClone;
  });

  let downloadButtonStatus = true;
  if (props.downloadType !== '') {
    downloadButtonStatus = false;
  }
  const selectedAttributes: any = getSelectedAttributes(props.attributes);
  const selectedFilters: any = getSelectedFilters(props.filters);

  return (
    <div className={styles.wrapper}>
      {!props.showPreview && (
        <div className={styles.resultCounter}>
          <span>{resultCount} results</span>
        </div>
      )}

      {props.showPreview && (
        <>
          <div className={styles.backButton}>
            <ImageButton
              onClick={showFilters}
              description={'Back'}
              image={BackIcon}
            />
          </div>
          <div className={styles.previewResultCounter}>
            <div>
              <span className={styles.boldResultCounter}>{resultCount}</span>
              <span className={styles.resultsLabel}>results</span>
            </div>
            <div className={styles.saveConfiguration}>Save configuration</div>
          </div>
        </>
      )}

      <div className={styles.selectedFilters}>
        <RoundButton
          onClick={filterOnClick}
          status={RoundButtonStatus.ACTIVE}
          classNames={styles}
        >
          {props.selectedPreFilter}
        </RoundButton>
        {props.showPreview && (
          <div className={styles.downloadTypeSelectHolder}>
            <span className={styles.downloadTypeLabel}>Download as </span>
            <span className={styles.downloadTypeSelect}>
              <Select
                options={selectOptions}
                onSelect={(option: string) => {
                  handleDownloadTypeSelect(option);
                }}
                placeholder={'Select'}
              />
            </span>
          </div>
        )}
      </div>

      <div className={styles.previewButton}>
        {!props.showPreview && (
          <SecondaryButton onClick={previewButtonOnClick}>
            Preview download
          </SecondaryButton>
        )}
        {props.showPreview && (
          <PrimaryButton
            isDisabled={downloadButtonStatus}
            onClick={() => {
              fetchCustomDownloadResults(
                props.downloadType,
                selectedAttributes,
                selectedFilters
              );
            }}
          >
            Download
          </PrimaryButton>
        )}
      </div>
    </div>
  );
};

type DispatchProps = {
  togglePreFiltersPanel: (togglePreFiltersPanel: boolean) => void;
  setShowPreview: (setShowPreview: boolean) => void;
  setDownloadType: (setDownloadType: string) => void;
};

const mapDispatchToProps: DispatchProps = {
  togglePreFiltersPanel,
  setShowPreview,
  setDownloadType
};

type StateProps = {
  selectedPreFilter: string;
  previewResult: any;
  showPreview: boolean;
  downloadType: string;
  filters: any;
  attributes: any;
};

const mapStateToProps = (state: RootState): StateProps => ({
  selectedPreFilter: getSelectedPreFilter(state),
  previewResult: getPreviewResult(state),
  showPreview: getShowPreviewResult(state),
  downloadType: getDownloadType(state),
  filters: getFilters(state),
  attributes: getAttributes(state)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Header);
