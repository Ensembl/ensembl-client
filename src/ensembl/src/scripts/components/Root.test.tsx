import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Root from './Root';
import Header from './layout/header/Header';
import Content from './layout/Content';

describe('<Root />', () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallow(<Root />);
  });

  describe('contains', () => {
    test('<Header />', () => {
      expect(wrapper.contains(<Header />)).toBe(true);
    });

    test('<Content />', () => {
      expect(wrapper.contains(<Content />)).toBe(true);
    });
  });

  test('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
