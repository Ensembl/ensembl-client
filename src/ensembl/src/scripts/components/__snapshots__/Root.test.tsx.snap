// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Root /> renders correctly 1`] = `
<BrowserRouter>
  <StrictMode>
    <Header />
    <withRouter(Connect(Content)) />
  </StrictMode>
</BrowserRouter>
`;
